driver/utils/unicode_conv.h

```diff
@@ -182,113 +196,84 @@ inline decltype(auto) toUTF8(const std::basic_string<CharType> & src) {
 template <>
-inline decltype(auto) fromUTF8<wchar_t>(const std::string & src) {
-    std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> convert;
-    return convert.from_bytes(src);
+inline decltype(auto) fromUTF8<char16_t>(const std::string & src, UnicodeConversionContext & context) {
+    return context.UCS2_converter_char16.from_bytes(src);
 }
 
 template <>
-inline decltype(auto) fromUTF8<signed char>(const std::string & src) {
-    const auto converted = fromUTF8<char>(src);
-    return std::basic_string<signed char>{converted.begin(), converted.end()};
+inline decltype(auto) fromUTF8<char32_t>(const std::string & src, UnicodeConversionContext & context) {
+    return context.UCS2_converter_char32.from_bytes(src);
 }
+#endif
 
 template <>
-inline decltype(auto) fromUTF8<unsigned char>(const std::string & src) {
-    const auto converted = fromUTF8<char>(src);
-    return std::basic_string<unsigned char>{converted.begin(), converted.end()};
+inline decltype(auto) fromUTF8<wchar_t>(const std::string & src, UnicodeConversionContext & context) {
+    return context.UCS2_converter_wchar.from_bytes(src);
 }
 
+#if !defined(_MSC_VER) || _MSC_VER >= 1920
 template <>
-inline decltype(auto) fromUTF8<unsigned short>(const std::string & src) {
+inline decltype(auto) fromUTF8<unsigned short>(const std::string & src, UnicodeConversionContext & context) {
     static_assert(sizeof(unsigned short) == sizeof(char16_t), "unsigned short doesn't match char16_t exactly");
-    const auto converted = fromUTF8<char16_t>(src);
+    const auto converted = fromUTF8<char16_t>(src, context);
     return std::basic_string<unsigned short>{converted.begin(), converted.end()};
 }
+#endif
 
 template <typename CharType>
-inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest, const std::locale& locale); // Leave unimplemented for general case.
-
-template <>
-inline void fromUTF8<char>(const std::string & src, std::string & dest, const std::locale& locale) {
-    dest = fromUTF8<char>(src, locale);
-}
-
-template <>
-inline void fromUTF8<signed char>(const std::string & src, std::basic_string<signed char> & dest, const std::locale& locale) {
-    dest = fromUTF8<signed char>(src, locale);
-}
-
-template <>
-inline void fromUTF8<unsigned char>(const std::string & src, std::basic_string<unsigned char> & dest, const std::locale& locale) {
-    dest = fromUTF8<unsigned char>(src, locale);
+inline decltype(auto) fromUTF8(const std::string & src) {
+    UnicodeConversionContext context;
+    return fromUTF8<CharType>(src, context);
 }
 
 template <typename CharType>
-inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest,
-    typename std::enable_if<std::is_assignable<
-        std::basic_string<CharType>, decltype(fromUTF8<CharType>(src))
-    >::value>::type * = nullptr
+inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest, UnicodeConversionContext & context,
```

Comment:
You completely lost me here, why do you need to check assignability of return type of one overload while you end up calling another overload?

Namelly, you SFINAE-check that this is well-formed: 
`dest = fromUTF8<CharType>(src)`
But three lines later you do:
`dest = fromUTF8<CharType>(src, context);`

Why ?

Reply:
There are cases when the type is is wider than it is required for storing wide chars, but the convention expect that the value is just stored there. So here I make sure I can assign one value to another and call the implemented correct conversion.

Reply:
I mean you check for one case and use another, so there is an implicit dependency: you expect that `fromUTF8<CharType>(src)` returns the same thing as `fromUTF8<CharType>(src, context)`.

Reply:
Ok, I see what you mean. This one is actually a complementary check, to the other. In this case, it it were the only case, I wouldn't need to do that.^ but since there is a separate specialization that deals with cases of non-assignable types, this one should cover everything that the other doesn't cover.